llvm-general/src/LLVM/General/Internal/FFI/OrcJITC.cpp

```diff
@@ -0,0 +1,134 @@
+#include "LLVM/General/Internal/FFI/OrcJIT.hpp"
+
+#include "llvm/ExecutionEngine/Orc/CompileUtils.h"
+#include "llvm/ExecutionEngine/Orc/LambdaResolver.h"
+
+#include "llvm/IR/Mangler.h"
+
+#include <type_traits>
+
+using namespace llvm;
+using namespace orc;
+
+static std::string mangle(StringRef name, LLVMTargetDataRef dataLayout) {
+    std::string mangledName;
+    {
+        raw_string_ostream mangledNameStream(mangledName);
+        Mangler::getNameWithPrefix(mangledNameStream, name,
+                                   *unwrap(dataLayout));
+    }
+    return mangledName;
+}
+
+LLVMIRCompileLayerRef
+LLVM_General_createIRCompileLayer(LLVMObjectLinkingLayerRef objectLayer,
+                                  LLVMTargetMachineRef tm) {
+    TargetMachine *tmm = unwrap(tm);
+    return new IRCompileLayer<ObjectLinkingLayer<>>(*objectLayer,
+                                                    SimpleCompiler(*tmm));
+}
+
+void LLVM_General_disposeIRCompileLayer(LLVMIRCompileLayerRef compileLayer) {
+    delete compileLayer;
+}
+
+LLVMJITSymbolRef LLVM_General_IRCompileLayer_findSymbol(
+    LLVMIRCompileLayerRef compileLayer, LLVMTargetDataRef dataLayout,
+    const char *name, LLVMBool exportedSymbolsOnly) {
+    JITSymbol symbol = compileLayer->findSymbol(name, exportedSymbolsOnly);
+    return new JITSymbol(symbol);
+}
+
+void LLVM_General_disposeJITSymbol(LLVMJITSymbolRef symbol) { delete symbol; }
+
+LLVMModuleSetHandleRef LLVM_General_IRCompileLayer_addModuleSet(
+    LLVMIRCompileLayerRef compileLayer, LLVMTargetDataRef dataLayout,
+    LLVMModuleRef *modules, unsigned moduleCount,
+    void (*dylibResolver)(const char *, LLVMJITSymbolRef),
+    void (*externalResolver)(const char *, LLVMJITSymbolRef)) {
+    auto lambdaResolver = createLambdaResolver(
+        [dylibResolver,
+         externalResolver](const std::string &name) -> RuntimeDyld::SymbolInfo {
+            JITSymbol symbol(nullptr);
+            dylibResolver(name.c_str(), &symbol);
+            return symbol.toRuntimeDyldSymbol();
+        },
+        [&](const std::string &name) -> RuntimeDyld::SymbolInfo {
+            JITSymbol symbol(nullptr);
+            externalResolver(name.c_str(), &symbol);
+            return symbol.toRuntimeDyldSymbol();
+        });
+    std::vector<Module *> moduleVec(moduleCount);
+    for (unsigned i = 0; i < moduleCount; ++i) {
+        moduleVec.at(i) = unwrap(modules[i]);
+        if (moduleVec.at(i)->getDataLayout().isDefault()) {
+            moduleVec.at(i)->setDataLayout(*unwrap(dataLayout));
+        }
+    }
+    return new IRCompileLayer<ObjectLinkingLayer<>>::ModuleSetHandleT(
```

Comment:
I’ve moved it into the encoding of `SymbolResolver` so we can easily reuse it for e.g. the lazy JIT. I don’t think exposing the encoding separately makes a lot of sense.